velox/exec/MergeJoin.cpp

```diff
@@ -1383,20 +1383,48 @@ RowVectorPtr MergeJoin::applyFilter(const RowVectorPtr& output) {
     // Before we leave the current buffer, since we may not have seen the next
     // left key match yet, the last key match may still be pending to produce
     // a row (because `processFilterResult()` was not called yet).
-    //
-    // To handle this, we need to call `noMoreFilterResults()` unless the
-    // same current left key match may continue in the next buffer. So there
-    // are two cases to check:
-    //
-    // 1. If leftMatch_ is nullopt, there for sure the next buffer will
-    // contain a different key match.
-    //
-    // 2. leftMatch_ may not be nullopt, but may be related to a different
-    // (subsequent) left key. So we check if the last row in the batch has the
-    // same left row number as the last key match.
-    if (!leftMatch_ || !joinTracker_->isCurrentLeftMatch(numRows - 1)) {
-      joinTracker_->noMoreFilterResults(onMiss);
-    }
+    auto onMissLastRow = [&](auto... args) {
+      // To handle this, we need to call `onMiss()` unless the
+      // same current left key match may continue in the next buffer.
+
+      // 1. If leftMatch_ is nullopt, there for sure the next buffer will
+      // contain a different key match.
+      if (!leftMatch_) {
+        onMiss(args...);
+        return;
+      }
+
+      // 2. leftMatch_ may not be nullopt, but may be related to a different
+      // (subsequent) left key. So we check if the last row in the batch has the
+      // same left row number as the last key match.
+      if (!joinTracker_->isCurrentLeftMatch(numRows - 1)) {
```

Comment:
Could we combine these two if conditions into one, like in the previous version of the code?